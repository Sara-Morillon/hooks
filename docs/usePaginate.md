# usePaginate

`usePaginate` offers an easy way to apply local pagination to a set of local data.

_:information_source: Pagination is 1-based._

## Example

```tsx
import { usePaginate } from '@saramorillon/hooks'

interface IData {
  name: string
  age: number
}

const data: IData[] = [
  { name: 'Zola Ray', age: 30 },
  { name: 'Arlo Curtis', age: 32 },
]

function MyComponent() {
  const { state, goTo, setLimit } = usePaginate<IData>(data)

  return (
    <>
      <table>
        <thead>
          <tr>
            <th>Name</th>
            <th>Age</th>
          </tr>
        </thead>

        <tbody>
          {rows.map((row) => (
            <tr>
              <td>{row.name}</td>
              <td>{row.age}</td>
            </tr>
          ))}
        </tbody>
      </table>

      <Pagination pagination={state.pagination} goTo={goTo} setLimit={setLimit} total={data.length} />
    </>
  )
}
```

## API

```typescript
usePaginate<T>(data: T[], initialPagination?: IPaginationInfo): IPaginated<T>
```

### Arguments

`data: T[]` - the data that will be paginated.

`initialPagination?: IPaginationInfo` - an object containing:

- `index: number` - the initial page index, default to 1.
- `limit: number` - the initial page size, default to 10.

### Returns

`IPaginated<T>` - an object containing:

- `state: IPaginationInfo` - an object containing:
  - `index: number` - the page index.
  - `limit: number` - the page size.
- `goTo: (index: number) => void` - a function to change page index.
- `setLimit: (limit: number) => void` - a function to change page size.
