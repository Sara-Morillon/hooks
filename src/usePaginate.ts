import { useCallback, useMemo, useState } from 'react'

export interface IPaginate {
  index: number
  limit: number
}

export function usePaginate<T>(data: T[]) {
  const [state, setState] = useState<IPaginate>({ index: 1, limit: 10 })

  const goTo = useCallback((index: number) => {
    setState((state) => ({ ...state, index }))
  }, [])

  const setLimit = useCallback((limit: number) => {
    setState((state) => ({ ...state, limit }))
  }, [])

  const rows = useMemo(() => {
    return data.slice((state.index - 1) * state.limit, state.index * state.limit)
  }, [data, state])

  return useMemo(() => ({ state, goTo, setLimit, rows }), [state, goTo, setLimit, rows])
}
