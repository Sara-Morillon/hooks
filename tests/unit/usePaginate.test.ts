import { act, renderHook } from '@testing-library/react'
import { usePaginate } from '../../src/usePaginate.js'
import { mockTableData } from '../mock.js'

describe('usePaginate', () => {
  it('should return first page rows by default', () => {
    const { result } = renderHook(() => usePaginate(mockTableData()))
    expect(result.current.rows).toEqual([
      { name: 'Zola Ray', age: 30 },
      { name: 'Arlo Curtis', age: 32 },
      { name: 'Alexis West', age: 39 },
      { name: 'Diego Fitzpatrick', age: 40 },
      { name: 'Annabella Vargas', age: 37 },
      { name: 'Ryker Gregory', age: 34 },
      { name: 'Alaya Winters', age: 40 },
      { name: 'Deandre Vu', age: 32 },
      { name: 'Kimora Higgins', age: 38 },
      { name: 'Sterling Pugh', age: 30 },
    ])
  })

  it('should return subset of rows when changing page index', () => {
    const { result } = renderHook(() => usePaginate(mockTableData()))
    act(() => result.current.goTo(2))
    expect(result.current.rows).toEqual([
      { name: 'Landry Pineda', age: 31 },
      { name: 'Gerardo Pearson', age: 39 },
      { name: 'Kiara Nava', age: 34 },
      { name: 'Stefan Hart', age: 35 },
      { name: 'Gemma Terrell', age: 38 },
      { name: 'Jaxen Smith', age: 35 },
      { name: 'Olivia Camacho', age: 30 },
    ])
  })

  it('should change subset size when changing limit', () => {
    const { result } = renderHook(() => usePaginate(mockTableData()))
    act(() => result.current.setLimit(5))
    expect(result.current.rows).toEqual([
      { name: 'Zola Ray', age: 30 },
      { name: 'Arlo Curtis', age: 32 },
      { name: 'Alexis West', age: 39 },
      { name: 'Diego Fitzpatrick', age: 40 },
      { name: 'Annabella Vargas', age: 37 },
    ])
  })
})
