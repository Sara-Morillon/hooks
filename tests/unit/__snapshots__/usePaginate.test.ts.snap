// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`usePaginate > should return subset of rows 1`] = `
[
  {
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "age": 35,
    "name": "Stefan Hart",
  },
]
`;

exports[`usePaginatedRows > should return subset of rows 1`] = `
[
  {
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "age": 35,
    "name": "Stefan Hart",
  },
]
`;
